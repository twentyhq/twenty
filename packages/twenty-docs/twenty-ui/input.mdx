---
title: Input
image: /images/user-guide/tips/light-bulb.png
---

<Frame>
  <img src="/images/user-guide/tips/light-bulb.png" alt="Header" />
</Frame>
