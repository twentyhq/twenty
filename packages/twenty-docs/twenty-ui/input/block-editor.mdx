---
title: Block Editor
image: /images/user-guide/api/api.png
---
<Frame>
  <img src="/images/user-guide/api/api.png" alt="Header" />
</Frame>

Uses a block-based rich text editor from [BlockNote](https://www.blocknotejs.org/) to allow users to edit and view blocks of content. 

<Tabs>
<Tab title="Usage">

```jsx
import { useBlockNote } from "@blocknote/react";
import { BlockEditor } from "@/ui/input/editor/components/BlockEditor";

export const MyComponent = () => {
  const BlockNoteEditor = useBlockNote();

  return <BlockEditor editor={BlockNoteEditor} />;
};
```

</Tab>
<Tab title="Props">


| Props | Type | Description |
|-------|------|-------------|
| editor | `BlockNoteEditor` | The block editor instance or configuration |



</Tab>
</Tabs>

