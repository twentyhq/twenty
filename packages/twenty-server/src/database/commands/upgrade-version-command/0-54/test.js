const scannedPositions = new Set();
let biggestPosition = -1;

// Sort options by position for consistent processing
const sortedOptions = [
  { name: 'a', position: 2 },
  { name: 'b', position: 1 },
  { name: 'c', position: 1 },
  { name: 'd', position: 2 },
].sort((a, b) => a.position - b.position);

for (const option of sortedOptions) {
  if (scannedPositions.has(option.position)) {
    option.position = biggestPosition + 1;
  }

  biggestPosition = Math.max(biggestPosition, option.position);
  scannedPositions.add(option.position);
}

console.log('Sorted options:', sortedOptions);
