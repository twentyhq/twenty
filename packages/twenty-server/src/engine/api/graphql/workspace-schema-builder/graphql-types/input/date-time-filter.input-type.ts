import { GraphQLISODateTime } from '@nestjs/graphql';

import { GraphQLInputObjectType, GraphQLList, GraphQLNonNull } from 'graphql';

import { FilterIs } from 'src/engine/api/graphql/workspace-schema-builder/graphql-types/input/filter-is.input-type';

export const DateTimeFilterType = new GraphQLInputObjectType({
  name: 'DateTimeFilter',
  fields: {
    eq: { type: GraphQLISODateTime },
    gt: { type: GraphQLISODateTime },
    gte: { type: GraphQLISODateTime },
    in: { type: new GraphQLList(new GraphQLNonNull(GraphQLISODateTime)) },
    lt: { type: GraphQLISODateTime },
    lte: { type: GraphQLISODateTime },
    neq: { type: GraphQLISODateTime },
    is: { type: FilterIs },
  },
});
