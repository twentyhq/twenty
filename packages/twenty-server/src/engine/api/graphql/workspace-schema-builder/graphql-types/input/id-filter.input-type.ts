import { GraphQLID, GraphQLInputObjectType, GraphQLList } from 'graphql';

import { FilterIs } from 'src/engine/api/graphql/workspace-schema-builder/graphql-types/input/filter-is.input-type';

export const IDFilterType = new GraphQLInputObjectType({
  name: 'IDFilter',
  fields: {
    eq: { type: GraphQLID },
    gt: { type: GraphQLID },
    gte: { type: GraphQLID },
    in: { type: new GraphQLList(GraphQLID) },
    lt: { type: GraphQLID },
    lte: { type: GraphQLID },
    neq: { type: GraphQLID },
    is: { type: FilterIs },
  },
});
