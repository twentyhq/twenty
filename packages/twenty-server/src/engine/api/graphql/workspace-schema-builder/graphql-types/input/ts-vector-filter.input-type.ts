import { GraphQLInputObjectType, GraphQLString } from 'graphql';

export const TSVectorFilterType = new GraphQLInputObjectType({
  name: 'TSVectorFilter',
  fields: {
    search: { type: GraphQLString },
  },
});
