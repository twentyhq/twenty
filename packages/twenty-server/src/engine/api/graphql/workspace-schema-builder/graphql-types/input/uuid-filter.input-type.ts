import { GraphQLInputObjectType, GraphQLList } from 'graphql';

import { FilterIs } from 'src/engine/api/graphql/workspace-schema-builder/graphql-types/input/filter-is.input-type';
import { UUIDScalarType } from 'src/engine/api/graphql/workspace-schema-builder/graphql-types/scalars';

export const UUIDFilterType = new GraphQLInputObjectType({
  name: 'UUIDFilter',
  fields: {
    eq: { type: UUIDScalarType },
    gt: { type: UUIDScalarType },
    gte: { type: UUIDScalarType },
    in: { type: new GraphQLList(UUIDScalarType) },
    lt: { type: UUIDScalarType },
    lte: { type: UUIDScalarType },
    neq: { type: UUIDScalarType },
    is: { type: FilterIs },
  },
});
