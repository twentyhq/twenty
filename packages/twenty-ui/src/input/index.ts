/*
 * _____                    _
 *|_   _|_      _____ _ __ | |_ _   _
 *  | | \ \ /\ / / _ \ '_ \| __| | | | Auto-generated file
 *  | |  \ V  V /  __/ | | | |_| |_| | Any edits to this will be overridden
 *  |_|   \_/\_/ \___|_| |_|\__|\__, |
 *                              |___/
 */

export type { AnimatedButtonProps } from './button/components/AnimatedButton';
export { AnimatedButton } from './button/components/AnimatedButton';
export type { AnimatedLightIconButtonProps } from './button/components/AnimatedLightIconButton';
export { AnimatedLightIconButton } from './button/components/AnimatedLightIconButton';
export type {
  ButtonSize,
  ButtonPosition,
  ButtonVariant,
  ButtonAccent,
  ButtonProps,
} from './button/components/Button/Button';
export { Button } from './button/components/Button/Button';
export { baseTransitionTiming } from './button/components/Button/constant';
export type { ButtonGroupProps } from './button/components/ButtonGroup';
export { ButtonGroup } from './button/components/ButtonGroup';
export { ColorPickerButton } from './button/components/ColorPickerButton';
export type {
  FloatingButtonSize,
  FloatingButtonPosition,
  FloatingButtonProps,
} from './button/components/FloatingButton';
export { FloatingButton } from './button/components/FloatingButton';
export type { FloatingButtonGroupProps } from './button/components/FloatingButtonGroup';
export { FloatingButtonGroup } from './button/components/FloatingButtonGroup';
export type {
  FloatingIconButtonSize,
  FloatingIconButtonPosition,
  FloatingIconButtonProps,
} from './button/components/FloatingIconButton';
export { FloatingIconButton } from './button/components/FloatingIconButton';
export type { FloatingIconButtonGroupProps } from './button/components/FloatingIconButtonGroup';
export { FloatingIconButtonGroup } from './button/components/FloatingIconButtonGroup';
export type {
  IconButtonSize,
  IconButtonPosition,
  IconButtonVariant,
  IconButtonAccent,
  IconButtonProps,
} from './button/components/IconButton';
export { IconButton } from './button/components/IconButton';
export type { IconButtonGroupProps } from './button/components/IconButtonGroup';
export { IconButtonGroup } from './button/components/IconButtonGroup';
export type {
  LightButtonAccent,
  LightButtonProps,
} from './button/components/LightButton';
export { LightButton } from './button/components/LightButton';
export type {
  LightIconButtonAccent,
  LightIconButtonSize,
  LightIconButtonProps,
} from './button/components/LightIconButton';
export { LightIconButton } from './button/components/LightIconButton';
export type { LightIconButtonGroupProps } from './button/components/LightIconButtonGroup';
export { LightIconButtonGroup } from './button/components/LightIconButtonGroup';
export type { MainButtonVariant } from './button/components/MainButton';
export { MainButton } from './button/components/MainButton';
export { RoundedIconButton } from './button/components/RoundedIconButton';
export { CodeEditor } from './code-editor/components/CodeEditor';
export type { CoreEditorHeaderProps } from './code-editor/components/CodeEditorHeader';
export { CoreEditorHeader } from './code-editor/components/CodeEditorHeader';
export { BASE_CODE_EDITOR_THEME_ID } from './code-editor/constants/BaseCodeEditorThemeId';
export { getBaseCodeEditorTheme } from './code-editor/theme/utils/getBaseCodeEditorTheme';
export type {
  ColorSchemeSegmentProps,
  ColorSchemeCardProps,
} from './color-scheme/components/ColorSchemeCard';
export { ColorSchemeCard } from './color-scheme/components/ColorSchemeCard';
export type { ColorSchemePickerProps } from './color-scheme/components/ColorSchemePicker';
export { ColorSchemePicker } from './color-scheme/components/ColorSchemePicker';
export { CardPicker } from './components/CardPicker';
export {
  CheckboxVariant,
  CheckboxShape,
  CheckboxSize,
  CheckboxAccent,
  Checkbox,
} from './components/Checkbox';
export { IconListViewGrip } from './components/IconListViewGrip';
export type { RadioProps } from './components/Radio';
export { RadioSize, LabelPosition, Radio } from './components/Radio';
export { RadioGroup } from './components/RadioGroup';
export type { ToggleSize, ToggleProps } from './components/Toggle';
export { Toggle } from './components/Toggle';
export type { ColorScheme } from './types/ColorScheme';
export type { SelectOption } from './types/SelectOption';
