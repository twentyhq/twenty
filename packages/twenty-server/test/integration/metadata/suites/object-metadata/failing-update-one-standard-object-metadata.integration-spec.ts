import { expectOneNotInternalServerErrorSnapshot } from 'test/integration/graphql/utils/expect-one-not-internal-server-error-snapshot.util';
import { findManyObjectMetadata } from 'test/integration/metadata/suites/object-metadata/utils/find-many-object-metadata.util';
import { updateOneObjectMetadata } from 'test/integration/metadata/suites/object-metadata/utils/update-one-object-metadata.util';
import { jestExpectToBeDefined } from 'test/utils/expect-to-be-defined.util.test';
import {
  eachTestingContextFilter,
  type EachTestingContext,
} from 'twenty-shared/testing';

import { type UpdateObjectPayload } from 'src/engine/metadata-modules/object-metadata/dtos/update-object.input';

type TestingRuntimeContext = {
  objectMetadataId: string;
  textFieldMetadataId: string;
};

type UpdateOneStandardObjectMetadataTestingContext = EachTestingContext<
  | ((args: TestingRuntimeContext) => Partial<UpdateObjectPayload>)
  | Partial<UpdateObjectPayload>
>[];

const failingUpdateTestsUseCase: UpdateOneStandardObjectMetadataTestingContext =
  [
    {
      title: 'when trying to update nameSingular on standard object',
      context: {
        nameSingular: 'newCompany',
      },
    },
    {
      title: 'when trying to update namePlural on standard object',
      context: {
        namePlural: 'newCompanies',
      },
    },
    {
      title:
        'when trying to update labelIdentifierFieldMetadataId on standard object',
      context: ({ textFieldMetadataId }) => ({
        labelIdentifierFieldMetadataId: textFieldMetadataId,
      }),
    },
    {
      title: 'when trying to update several forbidden properties',
      context: {
        namePlural: 'newCompanies',
        isActive: false,
        isLabelSyncedWithName: false,
        shortcut: 'whatever',
      },
    },
  ];

describe('Standard object metadata update should fail', () => {
  let companyObjectMetadataId: string;
  let companyNameFieldMetadataId: string;

  beforeAll(async () => {
    const { objects } = await findManyObjectMetadata({
      expectToFail: false,
      input: {
        filter: {},
        paging: { first: 100 },
      },
      gqlFields: `
        id
        nameSingular
        namePlural
        labelSingular
        labelPlural
        isLabelSyncedWithName
        fieldsList {
          id
          name
          type
        }
      `,
    });

    const companyObject = objects.find((o) => o.nameSingular === 'company');

    jestExpectToBeDefined(companyObject);
    companyObjectMetadataId = companyObject.id;

    const nameField = companyObject.fieldsList?.find(
      (field: { name: string }) => field.name === 'name',
    );

    jestExpectToBeDefined(nameField);
    companyNameFieldMetadataId = nameField!.id;
  });

  it.each(eachTestingContextFilter(failingUpdateTestsUseCase))(
    '$title',
    async ({ context }) => {
      const updatePayload =
        typeof context === 'function'
          ? context({
              objectMetadataId: companyObjectMetadataId,
              textFieldMetadataId: companyNameFieldMetadataId,
            })
          : context;

      const { errors } = await updateOneObjectMetadata({
        input: {
          idToUpdate: companyObjectMetadataId,
          updatePayload,
        },
        expectToFail: true,
      });

      expectOneNotInternalServerErrorSnapshot({ errors });
    },
  );
});
