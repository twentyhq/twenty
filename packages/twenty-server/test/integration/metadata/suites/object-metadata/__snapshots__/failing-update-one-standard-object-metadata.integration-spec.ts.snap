// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Standard object metadata update should fail when trying to update labelIdentifierFieldMetadataId on standard object 1`] = `
{
  "extensions": {
    "code": "BAD_USER_INPUT",
    "subCode": "INVALID_OBJECT_INPUT",
    "userFriendlyMessage": "An error occurred.",
  },
  "message": "Cannot edit standard object metadata properties: labelIdentifierFieldMetadataId",
  "name": "UserInputError",
}
`;

exports[`Standard object metadata update should fail when trying to update namePlural on standard object 1`] = `
{
  "extensions": {
    "code": "BAD_USER_INPUT",
    "subCode": "INVALID_OBJECT_INPUT",
    "userFriendlyMessage": "An error occurred.",
  },
  "message": "Cannot edit standard object metadata properties: namePlural",
  "name": "UserInputError",
}
`;

exports[`Standard object metadata update should fail when trying to update nameSingular on standard object 1`] = `
{
  "extensions": {
    "code": "BAD_USER_INPUT",
    "subCode": "INVALID_OBJECT_INPUT",
    "userFriendlyMessage": "An error occurred.",
  },
  "message": "Cannot edit standard object metadata properties: nameSingular",
  "name": "UserInputError",
}
`;

exports[`Standard object metadata update should fail when trying to update several forbidden properties 1`] = `
{
  "extensions": {
    "code": "BAD_USER_INPUT",
    "subCode": "INVALID_OBJECT_INPUT",
    "userFriendlyMessage": "An error occurred.",
  },
  "message": "Cannot edit standard object metadata properties: isLabelSyncedWithName, namePlural",
  "name": "UserInputError",
}
`;
